### ISSUE TITLE
Wrong gradient for torch.norm(x, p=float('inf')) when input tensor has non-unique max values

### ISSUE THREAD

**Issue by yku12cn** [pytorch/pytorch#41779]
## 🐛 Bug

Given the input tensor:
```
tensor([[ 9.,  2.,  9.],
        [-2., -3., -4.],
        [ 7.,  8., -9.]])
```
after norm(x, p=float('inf')), and backward(), the x.grad is:
```
tensor([[ 1.,  0.,  1.],
        [-0., -0., -0.],
        [ 0.,  0., -1.]])
```
While the correct grad should be:
```
tensor([[ 0.3333,  0.0000,  0.3333],
        [-0.0000, -0.0000, -0.0000],
        [ 0.0000,  0.0000, -0.3333]])
```

## To Reproduce
Run following code:
```
with torch.enable_grad():
    a = torch.tensor([
        [9., 2., 9.],
        [-2., -3., -4.],
        [7., 8., -9.],
    ], requires_grad=True)
    b = torch.norm(a, p=float('inf'))
    b.backward()
    print(a.grad)
```

## Expected behavior

```
tensor([[ 0.3333,  0.0000,  0.3333],
        [-0.0000, -0.0000, -0.0000],
        [ 0.0000,  0.0000, -0.3333]])
```

## Environment

PyTorch version: 1.5.0
Is debug build: No
CUDA used to build PyTorch: 10.2

OS: Microsoft Windows 10 Pro
GCC version: Could not collect
CMake version: Could not collect

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: Could not collect        
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect       
cuDNN version: Could not collect

Versions of relevant libraries:
[pip3] numpy==1.18.5
[pip3] torch==1.5.0
[pip3] torchvision==0.6.0
[conda] blas                      1.0                         mkl
[conda] cudatoolkit               10.2.89              h74a9793_0
[conda] mkl                       2020.1                      216
[conda] mkl-service               2.3.0            py37hb782905_0
[conda] mkl_fft                   1.1.0            py37h45dec08_0
[conda] mkl_random                1.1.1            py37h47e9c7a_0
[conda] numpy                     1.18.5           py37h6530119_0
[conda] numpy-base                1.18.5           py37hc3f5095_0
[conda] pytorch                   1.5.0           py3.7_cuda102_cudnn7_0    pytorch
[conda] torchvision               0.6.0                py37_cu102    pytorch

## Additional context

Mathematically speaking, the gradient of l_n norm should converge to the gradient of l_∞ while n grow.
with the following code
```
with torch.enable_grad():
    a = torch.tensor([
        [9., 2., 9.],
        [-2., -3., -4.],
        [7., 8., -9.],
    ], requires_grad=True)
    b = torch.norm(a, p=30)
    b.backward()
    print(a.grad)
```
Since we feed a rather large 'p', its output becomes:
```
tensor([[ 3.4248e-01,  3.9037e-20,  3.4248e-01],
        [-3.9037e-20, -4.9903e-15, -2.0958e-11],
        [ 2.3413e-04,  1.1252e-02, -3.4248e-01]])
```
which is much closer to
```
tensor([[ 0.3333,  0.0000,  0.3333],
        [-0.0000, -0.0000, -0.0000],
        [ 0.0000,  0.0000, -0.3333]])
```
rather than
```
tensor([[ 1.,  0.,  1.],
        [-0., -0., -0.],
        [ 0.,  0., -1.]])
```
In another word, '∂torch.norm(x, n)/∂x' does not converge to '∂torch.norm(x, float('inf'))/∂x' as n grow while there is non-unique max values in the input tensor.

cc @ezyang @SsnL @albanD @zou3519 @gqchen

**Comment by albanD** [comment]
Very good catch! Thanks for reporting it.
I think the subgradient we want here is `1` (or `-1`) on one of them and `0` for all the others. Just to be consistent with how we do max, maxpool, etc.

**Comment by yku12cn** [comment]
> Very good catch! Thanks for reporting it.
> I think the subgradient we want here is `1` (or `-1`) on one of them and `0` for all the others. Just to be consistent with how we do max, maxpool, etc.

@albanD  In the sense of training a NN by SGD, this bug won't make too much of differences. Especially when it's almost impossible to have two identical max value in the matrix. However, speaking of math and research. Thankfully I noticed it before it ruined my work.....

**Comment by albanD** [comment]
Let me give an example to be sure we talk about the same thing:
For input being [9, 0, -9], result is 9.
Current grad is [1, 0, -1] which is definitely bad!
Possible subgradients we could use are [0.5, 0, -0.5], [1,0,0] or [0,0,-1] (other things in the middle as well, but I don't see any good reason to consider them).

The main arguments for the first one are:
- It has the minimum 2-norm.
- It corresponds to the limit as p->inf
The main argument for the other two (any of them):
- It is consistent with what we would do if you did `inp.abs().max()`
- It is consistent with what we do for maxpool/max in general

**Comment by yku12cn** [comment]
> Let me give an example to be sure we talk about the same thing:
> For input being [9, 0, -9], result is 9.
> Current grad is [1, 0, -1] which is definitely bad!
> Possible subgradients we could use are [0.5, 0, -0.5], [1,0,0] or [0,0,-1] (other things in the middle as well, but I don't see any good reason to consider them).
> 
> The main arguments for the first one are:
> 
> * It has the minimum 2-norm.
> * It corresponds to the limit as p->inf
> 
> The main argument for the other two (any of them):
> 
> * It is consistent with what we would do if you did `inp.abs().max()`
> * It is consistent with what we do for maxpool/max in general

@albanD Thanks for your clarification. I would go with [0.5, 0, -0.5].
As for the gradient of max(), I wouldn't bother much, since there is not rigorous definition of how you calculate the max and it's not continuous. (Though I think the gradient of max([9, 4, 9]) should also be [0.5, 0, 0.5].... It makes more sense in math.)

The gradient of Lp norm, on the other hand, is very well [defined](https://en.wikipedia.org/wiki/Lp_space). Which means, any other results can not be justified mathematically. 
Another use case would be:
```
with torch.enable_grad():
    a = torch.tensor([
        [3., 0., 4.],
        [-1., -2., -4.],
        [-5., 0., 0.],
    ], requires_grad=True)
    b = torch.norm(a, p=2, dim=0)
    print(b)
    b = torch.norm(b, p=float('inf'))
    b.backward()
    print(a.grad)
```

**Comment by albanD** [comment]
> The gradient of Lp norm, on the other hand, is very well defined.

Can you point in that page where the derivative is well defined? The function itself is well defined. But at a point where multiple inputs are equal, it is not continuously differentiable. So at this point, any sub-differential can be chosen (as the function is convex). Am I missing something?

In any case, I think the min-norm argument is the one we want to follow here. So we should update [this](https://github.com/pytorch/pytorch/blob/dac393fa245aceaa95ec72b45801a0b90d1e8c6e/tools/autograd/templates/Functions.cpp#L126) formula to divide by the number of values equal to the max.

**Comment by yku12cn** [comment]
> > The gradient of Lp norm, on the other hand, is very well defined.
> 
> Can you point in that page where the derivative is well defined? The function itself is well defined. But at a point where multiple inputs are equal, it is not continuously differentiable. So at this point, any sub-differential can be chosen (as the function is convex). Am I missing something?
> 
> In any case, I think the min-norm argument is the one we want to follow here. So we should update [this](https://github.com/pytorch/pytorch/blob/dac393fa245aceaa95ec72b45801a0b90d1e8c6e/tools/autograd/templates/Functions.cpp#L126) formula to divide by the number of values equal to the max.

@albanD Given the definition of lp Norm, you can take the derivative and then take the limit of p-> inf.  IDK how to post formula here...The prove is rather simple.

**Comment by albanD** [comment]
Right, I think we agree on the result, just not on the exact naming.
The function `inf_norm(x)` is not actually continuously differentiable (in the strict sense) as depending on the direction in which you approach (different x coordinates) you get a different gradient. And so the limit used to define the existence of the derivative does not exist.
But for that function, you can define sub-gradients as it is convex.
And following the good practice (ensure that it is a descent direction) of using the min-norm subgradient, `1/nb_max` is a good choice.

Now if you consider `norm(x, p)`, here again, for p=inf, the function is not continuously differentiable when entries of x are equal. But you can define the gradient by continuity when p->inf as you show above. But again this is an arbitrary choice that we're making. You could also define it by continuity by having p=inf et having x->x0 (where x0 is the point with multiple entries that are equal) in that case, you would get one of the `[1, 0, 0]`-like solution above.


In any case, for convex functions, min-norm subgradients should be used. So no need to try and find an arbitrary definition by contiguity.

**Comment by yku12cn** [comment]
@albanD  May I know why it's good to choose a min-norm subgradients in practice?

**Comment by albanD** [comment]
> May I know why it's good to choose a min-norm subgradients in practice?

Because all subgradients might not be descent directions. For example. for absolute value at `0`, `1` is a valid subgradient. But it is not a good direction to follow.
But the minimum norm subgradient is always a descent direction. You can check "Convex Optimization Algorithms" by Dimitri Bertsekas, Proposition 4.3.1 for a proof.
